ONSdigital/dp-api-clients-go
# Stream cantabular CSV and add a checker func for Cantabular API Ext

### What

Stream CSV response from Cantabular so that we don't need to keep the whole response in memory.

Trello card: https://trello.com/c/Ph3rIO3M/5354-update-dp-cantabular-csv-exporter-to-use-latest-version-of-code-from-scc

SCC code reference
https://github.com/cantabular/examples/tree/master/golang/cmd/cantabular-query-streamed

- Implemented generic Stream function that uses an io.Pipe to read, transform, and write on the fly (used @redhug1  Cantabular XLSX implementation as a reference)
- Copied JSON streamer from SCC, added error handling, and supported a few extra corner cases for 'data' being decoded before 'error'
- Implemented a new function for Cantabular client, called `StaticDatasetQueryStreamCSV`, where the caller only needs to provide the Consumer method that will consume the CSV stream.
- Added unit tests

Required by [dp-cantabular-csv-exporter](https://github.com/ONSdigital/dp-cantabular-csv-exporter/pull/42)

### How to review

- Make sure code changes make sense
- Make sure the new method `StaticDatasetQueryStreamCSV` is easy to use by a caller and gives enough flexibility
- Make sure unit test pass
- (info only, not required - test with a real client):
  - You can start Cantabular server and api ext docker conainers
  - You can replace the unit-test mock with a real http client, and the unit tests should work as expected:
```
		cantabularClient := cantabular.NewClient(
			cantabular.Config{
				Host:           "http://localhost:8491",
				ExtApiHost:     "http://localhost:8492",
				GraphQLTimeout: 10 * time.Second,
			},
			dphttp.NewClient(),
			nil,
		)
```

### Who can review

anyone
